Gifski/Preview/PreBakedFrames.swift

```diff
@@ -0,0 +1,86 @@
+import Foundation
+import AVKit
+
+/**
+ Bug in AVPlayer: if the left trimmer is to the far left (the beginning of the video) or the far right, the avVideoPlayer will prevent frame redraws on swiftUI updates (like pressing the showPreviewButton on / off, or switching to crop mode). I have tried everything I can think of to find the real source of the error, but for now here is a workaround. We offset the track by just a bit on each end and prebake the frames just for the end. This offset will prevent the fullPreviewTrack from spanning the entire video range, and the pre-render keeps smooth video playback.
+ */
+struct PreBakedFrames {
+	private let frames: [Int: CGImage]
+	private let frameRate: Double
+
+	private let offsets: Offsets
+	private let timeRange: ClosedRange<Double>
+	private let numberOfFrames: Int
+
+	private static let bugFixOffset = 0.1
+
+	init(_ imageSource: CGImageSource, settings: SettingsForFullPreview) async throws {
+		numberOfFrames = CGImageSourceGetCount(imageSource)
+		guard numberOfFrames > 0 else {
+			throw CreateAVAssetError.noImages
+		}
+		frameRate = Double(settings.conversion.frameRate ?? 30)
+		offsets = Offsets(settings: settings)
+		timeRange = settings.conversion.timeRange ?? 0...settings.assetDuration
+		let speed = settings.speed
+
+		// +1 just to be safe
+		let numberOfFramesToPrecomputeAtTheBeginning = min(
+			numberOfFrames,
+			Int(ceil(offsets.start.seconds * frameRate * speed)) + 1
+		)
+		let numberOfFramesToPrecomputeAtTheEnd = min(
+			numberOfFrames,
+			Int(ceil(offsets.end.seconds * frameRate * speed)) + 1
+		)
+
+		let beginningRange = Set(0..<numberOfFramesToPrecomputeAtTheBeginning)
+		let endRange = Set(numberOfFrames - numberOfFramesToPrecomputeAtTheEnd..<numberOfFrames)
+		let allFrames = beginningRange.union(endRange)
+
+		let dict = try await allFrames.asyncMap { frameIndex in
```

Comment:
There is no benefit of using asyncMap here. It's not concurrent, it just allows waiting for asynchronous operations. Also, concurrency may be problematic as I'm not sure if CGImageSource is thread-safe.

```suggestion
		let dict = try await allFrames.map { frameIndex in
```

Reply:
Could maybe also use some cancellation support:

```swift
try Task.checkCancellation()
```